init.lua

```diff
@@ -33,111 +74,304 @@ local function get_int_attribute(player, key)
-	local name = player:get_player_name()
-	if not name then
+	if not is_player(player) then
 		return
 	end
 
-	local exhaustion = stamina_get_exhaustion(player) or 0
+	local exhaustion = stamina.get_exhaustion(player) or 0
 
-	exhaustion = exhaustion + v
+	exhaustion = exhaustion + change
 
-	if exhaustion > STAMINA_EXHAUST_LVL then
-		exhaustion = 0
-		local h = get_int_attribute(player, "stamina:level")
-		if h > 0 then
-			stamina_update_level(player, h - 1)
-		end
+	if exhaustion >= settings.exhaust_lvl then
+		exhaustion = exhaustion - settings.exhaust_lvl
+		stamina.change(player, -1)
 	end
 
-	player:set_attribute("stamina:exhaustion", exhaustion)
+	stamina.set_exhaustion(player, exhaustion)
+end
+--- END EXHAUSTION API ---
+--- SPRINTING API ---
+stamina.registered_on_sprintings = {}
+function stamina.register_on_sprinting(fun)
+	table.insert(stamina.registered_on_sprintings, fun)
 end
 
--- Sprint settings and function
-local enable_sprint = minetest.setting_getbool("sprint") ~= false
-local enable_sprint_particles = minetest.setting_getbool("sprint_particles") ~= false
-local armor_mod = minetest.get_modpath("3d_armor")
-
-function set_sprinting(name, sprinting)
-	local player = minetest.get_player_by_name(name)
-	local def = {}
-	-- Get player physics from 3d_armor mod
-	if armor_mod and armor and armor.def then
-		def.speed = armor.def[name].speed
-		def.jump = armor.def[name].jump
-		def.gravity = armor.def[name].gravity
+function stamina.set_sprinting(player, sprinting)
+	for _, fun in ipairs(stamina.registered_on_sprintings) do
+		local rv = fun(player, sprinting)
+		if rv == true then
+			return
+		end
 	end
 
-	def.speed = def.speed or 1
-	def.jump = def.jump or 1
-	def.gravity = def.gravity or 1
+	if player_monoids_mod then
+		if sprinting then
+			player_monoids.speed:add_change(player, 1 + settings.sprint_speed, "stamina:physics")
+			player_monoids.jump:add_change(player, 1 + settings.sprint_jump, "stamina:physics")
+		else
+			player_monoids.speed:del_change(player, "stamina:physics")
+			player_monoids.jump:del_change(player, "stamina:physics")
+		end
+	else
+		local def
+		if armor_mod then
+			-- Get player physics from 3d_armor mod
+			local name = player:get_player_name()
+			def = {
+				speed=armor.def[name].speed,
+				jump=armor.def[name].jump,
+				gravity=armor.def[name].gravity
+			}
+		else
+			def = {
+				speed=1,
+				jump=1,
+				gravity=1
+			}
+		end
 
-	if sprinting == true then
+		if sprinting then
+			def.speed = def.speed + settings.sprint_speed
+			def.jump = def.jump + settings.sprint_jump
+		end
 
-		def.speed = def.speed + SPRINT_SPEED
-		def.jump = def.jump + SPRINT_JUMP
+		player:set_physics_override(def)
 	end
 
-	player:set_physics_override({
-		speed = def.speed,
-		jump = def.jump,
-		gravity = def.gravity
-	})
+	if settings.sprint_particles and sprinting then
+		local pos = player:getpos()
+		local node = minetest.get_node({x = pos.x, y = pos.y - 1, z = pos.z})
+		if node.name ~= "air" then
```

Comment:
Should better apply for all `drawtype = "airlike"` and maybe for liquid nodes. Some servers use air nodes other than `air` for different realms.